Tighten fallback conditions for compiled optim

mlazos opened the pull request:
Since we now will support `capturable=False` when it's valid, narrow the eager fallback conditions to the cases where `compile` will fail. The lone case here is when the user deletes the capturable flag; `state_steps` are on cuda and `capturable` is `False`. Because a cuda tensor is not supported in the `value` kwarg for foreach ops this results in an error. 

The fallback wrapper is changed to check the device of `state_steps` if `capturable=False`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124544
* #125826
* __->__ #125825

mlazos committed:
Tighten fallback conditions for compiled optim

[ghstack-poisoned]

mlazos committed:
Update on "Tighten fallback conditions for compiled optim"




[ghstack-poisoned]

mlazos committed:
Update on "Tighten fallback conditions for compiled optim"

Since we now will support `capturable=False` when it's valid, narrow the eager fallback conditions to the cases where `compile` will fail. The lone case here is when the user deletes the capturable flag; `state_steps` are on cuda and `capturable` is `False`. Because a cuda tensor is not supported in the `value` kwarg for foreach ops this results in an error. 

The fallback wrapper is changed to check the device of `state_steps` if `capturable=False`.




[ghstack-poisoned]

janeyx99 commented on torch/optim/optimizer.py:
```diff
@@ -127,12 +127,36 @@ def _disable_dynamo_if_unsupported(single_tensor_fn=None):
-        def maybe_fallback(self, *args, **kwargs):
-            if is_compiling() and not kwargs.get("capturable", False):
-                return torch._disable_dynamo(func(self, *args, **kwargs))
+        def maybe_fallback(*args, **kwargs):
```
From my understanding of our discussion, the only two times where compile should continue compiling is if capturable=True or if params are on CPU. So the conditional I am expecting is:

if not capturable and not all params on CPU, then disable.

It may be even better to use the flip conditional to be clear:
```
if capturable or if params are all on CPU:
   don't disable
else:
   disable
```

mlazos replied:
We support mixed cpu/gpu (and the compiler no longer sets capturable=True in this case). Here's my perspective on our discussion - it established that 1) we don't always set capturable=True (only when safe) and 2) we will now trace the non-capturable path, avoiding item calls in the short term until the scalar translation layer is ready. 

The compiler is able to handle all of the cases eager can except if capturable=False and step is on GPU.

janeyx99 reviewed (approved):
Ah okay, so my suggestion is safer in that we are enabling dynamo for things we know absolutely should work, and yours is going off the idea that dynamo should really support all cases and we should only exclude specific configs that we know don't work.

They're approaching the same thing from different sides.

mlazos committed:
Update on "Tighten fallback conditions for compiled optim"

Since we now will support `capturable=False` when it's valid, narrow the eager fallback conditions to the cases where `compile` will fail. The lone case here is when the user deletes the capturable flag; `state_steps` are on cuda and `capturable` is `False`. Because a cuda tensor is not supported in the `value` kwarg for foreach ops this results in an error. 

The fallback wrapper is changed to check the device of `state_steps` if `capturable=False`.




[ghstack-poisoned]

mlazos committed:
Update on "Tighten fallback conditions for compiled optim"

Since we now will support `capturable=False` when it's valid, narrow the eager fallback conditions to the cases where `compile` will fail. The lone case here is when the user deletes the capturable flag; `state_steps` are on cuda and `capturable` is `False`. Because a cuda tensor is not supported in the `value` kwarg for foreach ops this results in an error. 

The fallback wrapper is changed to check the device of `state_steps` if `capturable=False`.




[ghstack-poisoned]
